### PR TITLE
Fix opening PDF files with special characters in their name

### PR DESCRIPTION
**Pull request by danxuliu**
[Before the move to the viewer](https://github.com/nextcloud/files_pdfviewer/pull/195) the URL of the PDF file to load was got (indirectly) [using `Files.getDownloadUrl()`](https://github.com/nextcloud/files_pdfviewer/blob/df9f2eb4339811156eb5be390847f71f597d43df/js/previewplugin.js#L134), which [encodes each section of the path as a URI component](https://github.com/nextcloud/server/blob/0dfdf3ee9978fdfd119acaa98b8747026d60599b/apps/files/js/files.js#L226-L234). The full URL, in turn, was [again encoded as a URI component](https://github.com/nextcloud/files_pdfviewer/blob/df9f2eb4339811156eb5be390847f71f597d43df/js/previewplugin.js#L56) to set the source of the iframe in which the PDF viewer is loaded. When [the PDF viewer parsed the URL](https://github.com/nextcloud/files_pdfviewer/blob/df9f2eb4339811156eb5be390847f71f597d43df/vendor/pdfjs/web/viewer.js#L1788-L1790) it [decoded it once](https://github.com/nextcloud/files_pdfviewer/blob/df9f2eb4339811156eb5be390847f71f597d43df/vendor/pdfjs/web/viewer.js#L3614-L3626), so each section of the path was still encoded as a URI component. 

After the move to the viewer the URL of the PDF file to load was got from the `davPath` property set by the viewer, which [uses the filename as is](https://github.com/nextcloud/viewer/blob/47a7d33dcf7f7445211265e7b68f17645c1fe10d/src/mixins/PreviewUrl.js#L52). The full URL was then encoded as a URI component, so when the PDF viewer parsed and decoded the URL the raw filename was used. In most cases it worked fine, but if the filename included some special character, like `?`, the file failed to load.

Now, instead of using the `davPath` property set by the viewer, each section of the dav path is encoded as a URI component. The encoding tries to not make any assumption on the format used by the viewer to avoid being coupled too much.

I do not know if this should be something to be fixed in the viewer instead (although images with "?" in their name load fine :shrug:). However, even in that case, this should fix the PDF viewer until the changes in the viewer are ready.

## How to test

- Open the Files app
- Upload a PDF file
- Rename it to `test?.pdf`
- Open the PDF file

### Result with this pull request

The file is opened in the viewer.

### Result without this pull request

The viewer opens, but the file fails to open.

### CODE REVIEW

File: src/views/PDFView.vue [hunk 1]
```diff
@@ -33,9 +33,32 @@ export default {
...
+					encodedDavPath += '/' + encodeURIComponent(pathSections[i])
+				}
+			}
+
+			if (hasScheme) {
+				encodedDavPath = pathSections[0] + '//' + pathSections[2] + encodedDavPath
+			}
+
+			return encodedDavPath
+		},
```
Comment (skjnldsv) [inline src/views/PDFView.vue:61]: The question is: shouldn't this be on Viewer? we have the same issue for other files

Comment (danxuliu) [reply src/views/PDFView.vue:61]: Yes, that question is in the pull request description...

**Review by MorrisJobke — APPROVED** [review]
Tested and works 👍

**Comment by rullzer** [comment]
SO go or no go?

**Comment by danxuliu** [comment]
I would say merge now and then adjust later if it is fixed in the viewer instead of the PDF viewer. Otherwise Nextcloud 21 could end shipping with the issue.

**Pull request merged**